### ISSUE TITLE
Blazor WebAssembly Support for GRPC

### ISSUE THREAD

**Issue by cnefan31** [FastEndpoints/FastEndpoints#711]
Hello, as FastEndpoints is an amazing lib(Remote feature the most) that I would like to use it everywhere in my projects, when I try to integrate it to Blazor webassembly, it seems not supporting brower runtime, and nowhere to find alternatives, I tried it in another way to achieve this by breaking the source code in **Fastendpoints.Messaging.Remote** and remove SocketHttpHander instead of HttpClientHandler who is supporting browser.

here is my demo repo **[FastEndpoints](https://github.com/cnefan31/FastEndpoints)**

In the counter page, you can see the action of "command remote execution".

I split **Fastendpoints.Messaging.Remote** into **Fastendpoints.Messaging.Remote.Browser** and **Fastendpoints.Messaging.Remote.Web**.

**Fastendpoints.Messaging.Remote.Browser** for blazor webassembly (using **Grpc.Net.Client.Web**), no Event sub/pub support as  webassembly does not has IHost( corrent me if I was wrong).
**Fastendpoints.Messaging.Remote.Web** for asp.net core host api (enable GrpcWeb in **Grpc.AspNetCore.Web**)

from my understanding, **Fastendpoints.Messaging.Remote** is a combination of GRPC client and server, but sometimes it only needs client or server, hoping GrpcWeb would be integrated in FastEndpoints originally.

Thanks in advance.

**Comment by dj-nitehawk** [comment]
we never considered the `Fastendpoints.Messaging.Remote` would want to be used in blazor wasm as the feature is mainly geared towards server to server communication. splitting it up would be a good idea. does everything work after you split it? all the tests are passing? if it is, would you like to submit a PR?

anyhow, your repo doesn't seem to exist. is it in private mode?

**Comment by cnefan31** [comment]
> we never considered the `Fastendpoints.Messaging.Remote` would want to be used in blazor wasm as the feature is mainly geared towards server to server communication. splitting it up would be a good idea. does everything work after you split it? all the tests are passing? if it is, would you like to submit a PR?
> 
> anyhow, your repo doesn't seem to exist. is it in private mode?

**Comment by cnefan31** [comment]
> make it public now.
as command-handler mechanism feels good to me, so I would like use remote instead of web api, so I want to introduce it into blazor wasm, in my repo, the server side should have the full features but the browser side does not, it only have command, the event(storage) things does not exist because async/task not support in current blazor-wasm implementation(from my understanding).

**Comment by dj-nitehawk** [comment]
will have a look at your repo when I get a chance. 
so, have you broken anything in the process the splitting the client and server parts?
do the tests still pass?
if nothing's broken after doing so, would you like to submit a PR with your changes?

**Comment by cnefan31** [comment]
PR submitted. [#714](https://github.com/FastEndpoints/FastEndpoints/pull/714)

**Comment by KateyBee** [comment]
@cnefan31 Thank you for figuring out how to support Blazor Wasm and the PR. However, since it was a bit messy and contained duplicated classes in multiple projects, I cleaned it up and went with a slightly different arrangement of packages. You can now use the package `FastEndpoints.Messaging.Remote.Core` in Wasm projects. While `FastEndpoints.Messaging.Remote` lib will require the full .NET runtime which contains both client + server portions. Please try the latest beta `v5.26.0.16-beta` and let me know if you encounter any issues.

**Comment by cnefan31** [comment]
@KateyBee  Thanks for your great work.
As Blazor Wasm  has different hosting mode, for client side ICommand should be registered with IServiceProvider directly, in file **FastEndpoints/Src/Messaging/Messaging.Remote.Core/Client /RemoteConnectionExtensions.cs** a new extension method needs to be added to register for IServiceProvider like :
``` csharp
    public static IServiceProvider MapRemote(this IServiceProvider provider, string remoteAddress, Action<RemoteConnection> r)
    {
        r(new(remoteAddress, provider));
        var logger = provider..GetRequiredService<ILogger<RemoteConnectionCore>>();
        logger.RemoteConfigured(remoteAddress, RemoteConnectionCore.RemoteMap.Count);
        return provider;
    }
```

**Comment by KateyBee** [comment]
Thanks, added that in `v5.26.0.19-beta` which should be available on Nuget soon.
Let me know if there's anything else.

**Comment by cnefan31** [comment]
@KateyBee I think we are still missing some extension method for ICommand 
``` razor
@page "/counter"

<PageTitle>Counter</PageTitle>

<h1>Counter</h1>

<p role="status">Current count: @currentCount</p>
<p role="status">Event Time: @timeStamp</p>
<button class="btn btn-primary" @onclick="IncrementCount">Click me</button>

@code {
    private int currentCount = 0;
    private string timeStamp = string.Empty;

    private async Task IncrementCount()
    {
        currentCount++;

        try
        {
            var result = await new CreateOrderCommand
                {
                    OrderId = 1001 + currentCount,
                    CustomerName = "Holly Simms"
                }
                .RemoteExecuteAsync(); // Here no extension (RemoteExecuteAsync) method for ICommand, not able to execute from remote end
            if (result is not null)
            {
                Console.WriteLine(result.Message);
                timeStamp = result.EventTime.ToString("yyyy-MM-dd HH:mm:ss.fff");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.ToString());
        }
    }
}
```
Some remote execute async methods like below(also copy from **FastEndpoints.Messaging.Remote**):

```csharp
    /// <summary>
    /// execute the command on the relevant remote server
    /// </summary>
    /// <param name="command"></param>
    /// <param name="options">call options</param>
    /// <exception cref="InvalidOperationException">thrown if the relevant remote handler has not been registered</exception>
    public static Task RemoteExecuteAsync(this ICommand command, CallOptions options = default)
    {
        var tCommand = command.GetType();

        return
            RemoteConnection.RemoteMap.TryGetValue(tCommand, out var remote)
                ? remote.ExecuteVoid(command, tCommand, options)
                : throw new InvalidOperationException($"No remote handler has been mapped for the command: [{tCommand.FullName}]");
    }

    /// <summary>
    /// execute the command on the relevant remote server and get back a result
    /// </summary>
    /// <typeparam name="TResult">the type of the result</typeparam>
    /// <param name="command"></param>
    /// <param name="options">call options</param>
    /// <exception cref="InvalidOperationException">thrown if the relevant remote handler has not been registered</exception>
    public static Task<TResult> RemoteExecuteAsync<TResult>(this ICommand<TResult> command, CallOptions options = default) where TResult : class
    {
        var tCommand = command.GetType();

        return
            RemoteConnection.RemoteMap.TryGetValue(tCommand, out var remote)
                ? remote.ExecuteUnary(command, tCommand, options)
                : throw new InvalidOperationException($"No remote handler has been mapped for the command: [{tCommand.FullName}]");
    }


    /// <summary>
    /// execute the command on the relevant remote server and get back a stream of <typeparamref name="TResult" />
    /// </summary>
    /// <typeparam name="TResult">the type of the result stream</typeparam>
    /// <param name="command"></param>
    /// <param name="options">call options</param>
    /// <exception cref="InvalidOperationException">thrown if the relevant remote handler has not been registered</exception>
    public static IAsyncEnumerable<TResult> RemoteExecuteAsync<TResult>(this IServerStreamCommand<TResult> command, CallOptions options = default) where TResult : class
    {
        var tCommand = command.GetType();

        return
            RemoteConnection.RemoteMap.TryGetValue(tCommand, out var remote)
                ? remote.ExecuteServerStream(command, tCommand, options).ReadAllAsync(options.CancellationToken)
                : throw new InvalidOperationException($"No remote handler has been mapped for the command: [{tCommand.FullName}]");
    }

    /// <summary>
    /// send the stream of <typeparamref name="T" /> to the relevant remote server and get back a result of <typeparamref name="TResult" />
    /// </summary>
    /// <typeparam name="T">the type of item in the stream</typeparam>
    /// <typeparam name="TResult">the type of the result that will be returned when the stream ends</typeparam>
    /// <param name="commands">the stream to send</param>
    /// <param name="options">call options</param>
    /// <exception cref="InvalidOperationException">thrown if the relevant remote handler has not been registered</exception>
    public static Task<TResult> RemoteExecuteAsync<T, TResult>(this IAsyncEnumerable<T> commands, CallOptions options = default)
        where T : class
        where TResult : class
    {
        var tCommand = typeof(IAsyncEnumerable<T>);

        return
            RemoteConnection.RemoteMap.TryGetValue(tCommand, out var remote)
                ? remote.ExecuteClientStream<T, TResult>(commands, tCommand, options)
                : throw new InvalidOperationException($"No remote handler has been mapped for the command: [{tCommand.FullName}]");
    }
```

Only ICommand extension methods need to be added, for the Event is not applicable, from my understanding Event needs long running Task in background which Blazor Wasm does not have in place yet, maybe it will have in the future, but not now.

**Comment by cnefan31** [comment]
And in **FastEndpoints.Messaging.Remote** also needs introduce **Grpc.AspNetCore.Web**. 
`<PackageReference Include="Grpc.AspNetCore.Web" Version="2.63.0" />`

To enable GrpcWeb for Wasm Client, in Registers, some more lines/codes needed.
`routeBuilder.MapGrpcService<,>().EnableGrpcWeb();`

```csharp
    public GrpcServiceEndpointConventionBuilder Register<TCommand, THandler>()
        where TCommand : class, ICommand
        where THandler : class, ICommandHandler<TCommand>
    {
        var tHandler = _routeBuilder.ServiceProvider.GetService<ICommandHandler<TCommand>>()?.GetType();

        if (tHandler is not null)
        {
            var tService = typeof(VoidHandlerExecutor<,>).MakeGenericType(typeof(TCommand), tHandler);
            var mapMethod = _mapGrpcMethodInfo.MakeGenericMethod(tService);

            //return (GrpcServiceEndpointConventionBuilder)mapMethod.Invoke(null, [_routeBuilder])!;
            // change like below, this should be consider carefully, not sure what's this method for indicating `tHandler'.
            var builder  = (GrpcServiceEndpointConventionBuilder)mapMethod.Invoke(null, [_routeBuilder])!;
            builder.EnableGrpcWeb();
            return builder;
        }

        return _routeBuilder.MapGrpcService<VoidHandlerExecutor<TCommand, THandler>>().EnableGrpcWeb(); // add .EnableGrpcWeb();
    }
```
For the rest registeres methods, `(GrpcServiceEndpointConventionBuilder)`.EnableGrpcWeb()

**Comment by KateyBee** [comment]
There's no need to reference `Grpc.AspNetCore.Web` in `FastEndpoints.Messaging.Remote`. Check out the following project. You can just add it in your `Server` project and enable the GrpcWeb middleware:

https://github.com/FastEndpoints/Blazor-Wasm-Remote-Messaging-Demo/blob/bd84b6255139d2f1466933772e9db3b50d655637/Server/Program.cs#L13-L18

**Comment by cnefan31** [comment]
Hello @KateyBee , I found that when I tried to use the original Remote.Client, it came to use the `RemoteConnectionCore` to do the connection instead of `RemoteConnection` which will lead to an error `Status(StatusCode="Internal", Detail="Content-Type 'application/grpc-web-text' is not supported.")` whose backend does not use grpc-web actually.

As `FastEndpoints.Messaging.Remote.Core`  was reffered in`FastEndpoints.Messaging.Remote`, the `RemoteExecuteAsync` will call `RemoteConnectionCore` to do the execution, I think some duplicate codes/files are needed for client in `FastEndpoints.Messaging.Remote` instead of `FastEndpoints.Messaging.Remote.Core`.

the below lines state the entry of remote call:
```csharp
            CreateOrderCommand remoteCommand = new CreateOrderCommand();
            try
            {
                var result = await remoteCommand.RemoteExecuteAsync();
            }
            catch (Exception ex) // ERROR:  Status(StatusCode="Internal", Detail="Content-Type 'application/grpc-web-text' is not supported.
            {

            }
```
The following lines are how the remote call executed, not as expected using `RemoteConnection`
```csharp
    public static Task<TResult> RemoteExecuteAsync<TResult>(this ICommand<TResult> command, CallOptions options = default(CallOptions)) where TResult : class
    {
        Type type = command.GetType();
        if (!RemoteConnectionCore.RemoteMap.TryGetValue(type, out RemoteConnectionCore value))
        {
            throw new InvalidOperationException("No remote handler has been mapped for the command: [" + type.FullName + "]");
        }

        return value.ExecuteUnary(command, type, options);
    }
```

**Comment by KateyBee** [comment]
Microsoft docs say to [do this](https://learn.microsoft.com/en-us/aspnet/core/grpc/grpcweb?view=aspnetcore-8.0#http-protocol).

**Comment by cnefan31** [comment]
@KateyBee  Please check my new repo [FastRemoteDemo](https://github.com/cnefan31/FastRemoteDemo) regarding this, call from client to server in one application with the following kestrel setting but not reffered to `Grpc.AspNetCore.Web`, this issue will come up, but when I add `Grpc.AspNetCore.Web` and enable `GrpcWeb` by adding `app.UseGrpcWeb(new() { DefaultEnabled = true });`, remote call will success. it seems all the throughput is handled with  HttpHandler instead of the orignal SocketHttpHandler, please correct me if I got any misunderstanding.

```json 
    "Kestrel": {
        "Endpoints": {
            "grpc": {
                "Url": "http://*:8080",
                "Protocols": "Http2"
            }
        }
    }
```

**Comment by cnefan31** [comment]
@KateyBee  Please check my new repo [FastRemoteDemo](https://github.com/cnefan31/FastRemoteDemo) regarding this, call from client to server in one application with the following kestrel setting but not reffered to `Grpc.AspNetCore.Web`, this issue will come up, but when I add `Grpc.AspNetCore.Web` and enable `GrpcWeb` by adding `app.UseGrpcWeb(new() { DefaultEnabled = true });`, remote call will success. it seems all the throughput is handled with  HttpHandler instead of the orignal SocketHttpHandler, please correct me if I got any misunderstanding.

```json 
    "Kestrel": {
        "Endpoints": {
            "grpc": {
                "Url": "http://*:8080",
                "Protocols": "Http2"
            }
        }
    }
```

**Comment by dj-nitehawk** [comment]
Good catch!
Fixed in `v5.26.0.24-beta`.
From now on "Grpc Web" projects will have to call `.MapRemoteCore()` as `.MapRemote()` has been moved to the `FastEndpoints.Messaging.Remote` project.

**Comment by cnefan31** [comment]
Great!
Demo runs as expected, happy coding...

**Comment by cnefan31** [comment]
I have added Event to browser remote client in my local [FastEndpointsDemo](https://github.com/cnefan31/FastEndpointsDemo) with project `https://github.com/cnefan31/FastEndpointsDemo/tree/master/src/Messaging.Remote.Browser`, which I missed before as I thought Event would not be applicable for browser(multi-threading), I misunderstood the Remote Event thing.

This will make me easy to implement `server to client subscription` using server streaming instead of websocket (another way). AMAZING library!!!
I would like to do some web GUI development with `Blazor webassembly` and `FastEndpoints` using this remote messaging Mechanism.

**Comment by dj-nitehawk** [comment]
@cnefan31 
starting from `v5.27.0.11-beta`, browser client can now subscribe to events from grpc server.
[example here](https://github.com/FastEndpoints/Blazor-Wasm-Remote-Messaging-Demo/blob/main/Client/Program.cs#L17).

**Comment by cnefan31** [comment]
@dj-nitehawk 
hello, still blazor wasm issue, the same demo repo as before, when I open 2 pages while subscribe the same event, and server is "Broadcast", only one client can receive this event like round robin, not all, as you can see below:
even I set `h.RegisterEventHub<SomethingHappened>(HubMode.EventBroker);` 
or ``h.RegisterEventHub<SomethingHappened>(HubMode.EventPublisher);
not sure if it differs between grpc.web and normal grpc.
![image](https://github.com/user-attachments/assets/fcf24dda-5b23-45a6-8188-f40decc14456)

**Comment by dj-nitehawk** [comment]
i think you'll have to manage receiving data with some sort of shared service or central location and dispatch the received data to all opened pages or something (maybe via a static variable or di container singleton). not sure exactly. haven't worked with blazor wasm enough to give good advise.

**Comment by cnefan31** [comment]
even I use console application to subscribe the events from server, it still process in round robin with `h.RegisterEventHub<SomethingHappened>(HubMode.EventPublisher);`

**Comment by dj-nitehawk** [comment]
found the issue. fixed in latest beta. not sure how this will work in a web browser though.

https://github.com/FastEndpoints/FastEndpoints/blob/ea91b2541fe0a5a6d45b1ad161474e381f4265a6/Src/Library/changelog.md?plain=1#L104-L120

**Comment by cnefan31** [comment]
thanks, with unique clientIdentifier, both wasm and normal client can receive each event.

another thing I found is that the event delay, when the second event produced after more than 60s, it will wait until next event to trigger, not "realtime". 

due to lines below:

`await Task.WhenAny(subscriber.Sem.WaitAsync(cts.Token), Task.Delay(60000));`

not sure if it has some other performance consideration, but seems not necessary.
as I changed it like below, each event received as expected, would this be configurable or remove in FE?
`await Task.WhenAny(subscriber.Sem.WaitAsync(cts.Token));`
